cranelift-codegen/src/regalloc/branch_splitting.rs

```diff
@@ -0,0 +1,223 @@
+            // block.
+            self.cur.goto_last_inst(ebb);
+            let terminator_inst = self
+                .cur
+                .current_inst()
+                .expect("missing terminator instruction");
+            match self.cur.prev_inst() {
+                Some(inst) => {
+                    let opcode = self.cur.func.dfg[inst].opcode();
+                    if opcode.is_branch() {
+                        self.visit_conditional_branch(inst, opcode);
+                    } else {
+                        continue;
+                    }
+                }
+                // If there is only a terminator instruction, then there is no need to create a new
+                // block.
+                None => continue,
+            };
+            self.cur.goto_inst(terminator_inst);
+            self.visit_terminator_branch(terminator_inst);
+        }
+
+        // If blocks were added the cfg and domtree are inconsistent and must be recomputed.
+        if self.new_blocks {
+            self.cfg.compute(&self.cur.func);
+            self.domtree.compute(&self.cur.func, self.cfg);
+        }
+    }
+
+    fn visit_conditional_branch(&mut self, inst: Inst, opcode: Opcode) {
+        let target = match self.cur.func.dfg[inst] {
+            InstructionData::Branch { destination, .. }
+            | InstructionData::BranchIcmp { destination, .. }
+            | InstructionData::BranchInt { destination, .. }
+            | InstructionData::BranchFloat { destination, .. } => destination,
+            _ => panic!("Unexpected instruction in classify_branch"),
+        };
+
+        // If there are any parameters, split the edge.
+        if self.should_split_edge(inst, target) {
+            // Remember the branch arguments.
+            let jump_args: Vec<Value> = self
+                .cur
+                .func
+                .dfg
+                .inst_variable_args(inst)
+                .iter()
+                .map(|x| *x)
+                .collect();
+
+            // Create the block the branch will jump to.
+            let new_ebb = self.make_empty_ebb();
+
+            // Remove the arguments from the branch and make it jump to the new block.
+            match opcode {
+                Opcode::Brz => {
+                    let val = self.cur.func.dfg.inst_args(inst)[0];
+                    self.cur.func.dfg.replace(inst).brz(val, new_ebb, &[]);
+                }
+                Opcode::Brnz => {
+                    let val = self.cur.func.dfg.inst_args(inst)[0];
+                    self.cur.func.dfg.replace(inst).brnz(val, new_ebb, &[]);
+                }
+                Opcode::BrIcmp => {
+                    if let InstructionData::BranchIcmp { cond, .. } = self.cur.func.dfg[inst] {
+                        let x = self.cur.func.dfg.inst_args(inst)[0];
+                        let y = self.cur.func.dfg.inst_args(inst)[0];
```

Comment:
I'm not sure why this is correct. It looks suspicous. Might be worth a comment?

Reply:
I suspect this should be 1 in the second line, yeah.